presto-raptorx/src/main/java/com/facebook/presto/raptorx/storage/ChunkRecoveryManager.java

```diff
@@ -0,0 +1,427 @@
+
+        log.info("Copying chunk %s from chunk store...", chunkId);
+        long start = System.nanoTime();
+
+        try {
+            chunkStore.getChunk(chunkId, stagingFile);
+        }
+        catch (PrestoException e) {
+            stats.incrementChunkRecoveryFailure();
+            tryDeleteFile(stagingFile);
+            throw e;
+        }
+
+        // record transfer rate
+        Duration duration = nanosSince(start);
+        DataSize size = succinctBytes(stagingFile.length());
+        DataSize rate = dataRate(size, duration);
+        stats.addChunkRecoveryDataRate(rate, size, duration);
+
+        log.info("Copied chunk %s from chunk store in %s (%s at %s/s)", chunkId, duration, size, rate);
+
+        // check if copied file is valid
+        if (isFileCorrupt(stagingFile, chunkSize, chunkXxhash64)) {
+            stats.incrementChunkRecoveryFailure();
+            stats.incrementCorruptRecoveredFile();
+            quarantineFile(chunkId, stagingFile, "File is corrupt after recovery.");
+            throw new PrestoException(RAPTOR_CHUNKSTORE_CORRUPTION, "Chunk is corrupt in chunk store: " + chunkId);
+        }
+
+        // move to final location
+        try {
+            Files.move(stagingFile.toPath(), storageFile.toPath(), ATOMIC_MOVE);
+        }
+        catch (FileAlreadyExistsException e) {
+            // someone else already created it (should not happen, but safe to ignore)
+        }
+        catch (IOException e) {
+            stats.incrementChunkRecoveryFailure();
+            throw new PrestoException(RAPTOR_RECOVERY_ERROR, "Failed to move chunk: " + chunkId, e);
+        }
+        finally {
+            tryDeleteFile(stagingFile);
+        }
+
+        stats.incrementChunkRecoverySuccess();
+    }
+
+    private void quarantineFile(long chunkId, File file, String message)
+    {
+        File quarantine = storageService.getQuarantineFile(chunkId);
+        if (quarantine.exists()) {
+            log.error("%s Quarantine already exists: %s", message, quarantine);
+            return;
+        }
+
+        log.error("%s Quarantining corrupt file: %s", message, quarantine);
+        try {
+            Files.move(file.toPath(), quarantine.toPath(), ATOMIC_MOVE);
+        }
+        catch (IOException e) {
+            log.warn(e, "Quarantine of corrupt file failed");
+            tryDeleteFile(file);
+        }
+    }
+
+    private static void deleteFile(Path file)
+    {
+        try {
+            Files.deleteIfExists(file);
+        }
+        catch (IOException e) {
+            throw new PrestoException(RAPTOR_STORAGE_ERROR, "Failed to delete local file: " + file, e);
+        }
+    }
+
+    @SuppressWarnings("ResultOfMethodCallIgnored")
+    private static void tryDeleteFile(File file)
+    {
+        file.delete();
+    }
+
+    private static File temporarySuffix(File file)
+    {
+        return new File(file.getPath() + ".tmp-" + randomUUID());
```

Comment:
What if we add `raptor` somewhere here so if this ends up having a bug where we don't delete files, we more quickly get the attribution?

Reply:
This is a good idea if we were creating temporary files in the system temporary directory, but there's no need here since is a private method used to create a temporary file inside the Raptor staging directory.
```
File stagingFile = temporarySuffix(storageService.getStagingFile(chunkId));